hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/SyntheticIds.java

```diff
@@ -0,0 +1,115 @@
+ */
+@Singleton
+public class SyntheticIds {
+    private static final AccountID DEBIT_NON_CANONICAL_REFERENCE_ID =
+            AccountID.newBuilder().accountNum(0L).build();
+    private static final AccountID CREDIT_NON_CANONICAL_REFERENCE_ID =
+            AccountID.newBuilder().alias(Bytes.wrap(new byte[20])).build();
+
+    @Inject
+    public SyntheticIds() {
+        // Dagger2
+    }
+
+    /**
+     * Given a native operations, returns the {@link AddressIdConverter} to use for converting
+     * addresses to ids in this context.
+     *
+     * @param nativeOperations the native operations
+     * @return the converter
+     */
+    public AddressIdConverter converterFor(@NonNull final HederaNativeOperations nativeOperations) {
+        return new AddressIdConverter() {
+            @Override
+            public @NonNull AccountID convert(@NonNull final Address address) {
+                return syntheticIdFor(address, nativeOperations);
+            }
+
+            @Override
+            public @NonNull AccountID convertCredit(@NonNull Address address) {
+                return syntheticIdForCredit(address, nativeOperations);
+            }
+        };
+    }
+
+    private static @NonNull AccountID syntheticIdFor(
+            @NonNull final Address address, @NonNull final HederaNativeOperations nativeOperations) {
+        return internalSyntheticId(false, address, nativeOperations);
+    }
+
+    private static @NonNull AccountID syntheticIdForCredit(
+            @NonNull final Address address, @NonNull final HederaNativeOperations nativeOperations) {
+        return internalSyntheticId(true, address, nativeOperations);
+    }
+
+    private static @NonNull AccountID internalSyntheticId(
+            final boolean isCredit,
+            @NonNull final Address address,
+            @NonNull final HederaNativeOperations nativeOperations) {
+        requireNonNull(address);
+        final var accountNum = accountNumberForEvmReference(address, nativeOperations);
+        if (accountNum == MISSING_ENTITY_NUMBER) {
+            final var explicit = explicitFromHeadlong(address);
+            if (isLongZeroAddress(explicit)) {
+                // References to missing long-zero addresses are synthesized as aliases for
+                // credits and numeric ids for debits
+                return isCredit ? aliasIdWith(explicit) : numericIdWith(numberOfLongZero(explicit));
+            } else {
+                // References to missing EVM addresses are always synthesized as alias ids
+                return aliasIdWith(explicit);
+            }
+        } else if (accountNum == NON_CANONICAL_REFERENCE_NUMBER) {
+            // Non-canonical references result are synthesized as ids of the zero address,
+            // using a numeric id for a debit and an alias id for a credit
+            return isCredit ? CREDIT_NON_CANONICAL_REFERENCE_ID : DEBIT_NON_CANONICAL_REFERENCE_ID;
```

Comment:
I wonder if you can explain this?   Why do we need to use an alias id for credit?

Reply:
Just to match mono-service (which is a bit of an exercise in futility, but doable enough in spots like this 😵‍💫 .)

No real reason otherwise!